### DISCUSSION TITLE
Kubernetes 질문 드려요!

### DISCUSSION THREAD

**Discussion by Tyler-devop** [yoonjeong-kwon/fastcampus-kubernetes#5] [Q&A] [answered]
안녕하세요, 강의 만드시느라 고생많으셨습니다!
강의 듣고 현업관련 궁금한 사항들이 생겨 질문드립니다!

1. 구성시 오브젝트 실행 순서는 상관없나요? 보통 변경이 많은 오브젝트는 무엇인가요?
   
2. 현업에서 오브젝트는 어떻게 관리하나요? 폴더 tree 형식의 예시를 보여주실 수 있으신가요? 
   
3. Rolling update or Blue/Green or Canary 중 가장 많이 사용되는 배포방법은 어떤것인가요? (안정성이나 속도 고려해서)
    3-1) Blue/Green 배포 시 중간 배포시 잠깐의 서비스 끊김이 발생하지는 않나요?

4. 기업에서는 instance 에 kuber 설치해서 사용 or container service 사용 중 어떤것을 선호 하나요? 어떻게 변경되어가고 있는 추세인가요?

**Comment by yoonjeong-kwon** [comment, marked as answer]
안녕하세요. @Tyler-devop 님, 좋은 질문 감사합니다!

> 1. 구성시 오브젝트 실행 순서는 상관없나요? 보통 변경이 많은 오브젝트는 무엇인가요?

**오브젝트 실행 순서**

Deployment와 Service 오브젝트 간에 실행 순서는 동작에 영향을 주지 않습니다. 아래 그림에서처럼 Deployment를 생성하면 ReplicaSet과 Pod를 프로비저닝합니다. Service를 생성하면 우리가 선언한 파드가 있을 때만 파드들에 대한 Endpoint를 자동으로 만들고 트래픽을 그 Enpoint로 라우팅합니다. 클러스터 상에 생성한 파드가 없다면, 파드가 생성 되었을 때 쿠버네티스가 그 파드들의 Endpoint를 추가합니다. 
이처럼 쿠버네티스가 API Server로 들어온 생성 요청과 현재 쿠버네티스 클러스터 상에 존재하는 리소스 상태를 비교하여 작업을 수행하기 때문에 우리는 순서를 신경쓰지 않고도 작업을 완료할 수 있습니다. 
<img width="921" alt="image" src="https://user-images.githubusercontent.com/97735995/162486870-cfce8bdc-d39a-46ba-9f43-f039947d8292.png">
다만, ReplicaSet, Pod를 각각 직접 생성하시는 경우에는 레이블 셀렉터 설정에 따라 의도하지 않은 Pod가 선택될 수 있기 때문에 실행 순서와 파드 레이블 등을 신경쓰셔야 합니다. 현업에서는 ReplicaSet, Pod를 각각 생성하지 않고 Deployment를 이용해서 ReplicaSet과 Pod를 생성합니다.

**변경이 많은 오브젝트**

질문주신 '변경' 이라는 의미를 'yaml을 변경하는 빈도'로 정의하고 설명드리겠습니다. 가장 변경이 많은 오브젝트는 Deployment 입니다. 왜냐하면 소스코드를 변경하고 테스트하고자 할 때 혹은 새로운 버전을 운영에 릴리스할 때마다 도커 이미지가 변경될 것이고 그에 따라 Deployment에 선언해야하는 이미지 태그도 달라지기 때문입니다.
이와 달리 Service는 사내 (마이크로)서비스를 구분하기 위한 이름으로 사용하기 때문에 거의 변경이 일어나지 않습니다. 예를 들어 A팀의 order 서비스에서 B팀의 payment 서비스 API를 호출한다고 가정할 때, B팀에서 payment 파드들에 대한 Service 이름을 어떻게 지었느냐에 따라 A팀이 사용해야할 payment 서비스 호스트네임이 달라지게 됩니다. 따라서 호스트네임 성격의 Service 이름은 자주 변경할 수 없게 되죠. 
어떤 오브젝트를 어떻게 관리할 것인가는 회사 정책마다 달라지겠지만, 일반적으로 사내 서비스 이름이나 호스트 네이밍 규칙은 컨벤션으로 정의되어 있기 때문에 Service 오브젝트나 Ingress와 같이 네트워킹과 관련된 오브젝트는 미리 만들어두고 사용합니다.


> 2. 현업에서 오브젝트는 어떻게 관리하나요? 폴더 tree 형식의 예시를 보여주실 수 있으신가요?

이부분은 특별한 노하우가 필요한 영역은 아닌 것 같습니다. 회사나 팀, 서비스 규모에 따라 적절한 폴더 레벨을 구성 해볼 수 있을텐데요.
일반적으로 생각해볼 수 있는 구조는 오브젝트별로 yaml 파일을 구분해서 관리하거나 담당하는 서비스가 많다면 서비스별로 폴더를 나누어 관리할 수도 있습니다. 


> 3. Rolling update or Blue/Green or Canary 중 가장 많이 사용되는 배포방법은 어떤것인가요? (안정성이나 속도 고려해서)
>     3-1) Blue/Green 배포 시 중간 배포시 잠깐의 서비스 끊김이 발생하지는 않나요?

**배포시 잠깐의 서비스 끊김**

어떤 배포 방식이던 시간 차에 따른 일시적인 순단은 발생할 수 있습니다. 정상적인 배포가 이뤄지는 환경이라면 이정도의 순단은 문제가 되지 않습니다. 애플리케이션 레벨에서 일시적인 에러에 대한 fallback 처리가 제공되기 때문입니다. 물론 쿠버네티스에서 지원하는 배포 전략은 구글이 수년간 운영 경험 끝에 다운타임 발생을 최소화하기 위한 노하우를 담았기 때문에 과거 배포 방식(프로세스 셧다운 후 재기동 등)에 비해 안정적이죠. 

**가장 많이 사용되는 배포방법**

가장 많이 사용되는 배포방법을 하나로 말씀드리기는 어려울 것 같습니다. 각 배포 전략별로 특징이 다르기 때문에 상황에 따라 적절한 방법을 채택하셔야 합니다. Blue/Green의 경우 두배의 리소스가 필요하기 때문에 리소스가 불충분한 환경이라면 사용할 수 없을 것입니다. Canary의 경우, 새로 릴리스할 버전을 소수만 배포해보고 잠시동안 미리 기능을 테스트 해볼 수 있기 때문에 리스크가 예상되는 경우 적합한 전략이 될 수 있습니다. 만약 팀 내에서 Blue/Green을 이용해 모든 트래픽을 일시에 새로운 기능으로 전환하는 것이 부담 혹은 리스크가 크다고 판단된다면, Canary가 도움이 될 것입니다. 그러나 Canary는 같은 시점에 서로 다른 버전이 배포되어 있는 환경이기 때문에 어떤 사용자에게는 A 화면이, 다른 사용자에게는 B 화면이 노출될 수 있습니다. 이것이 문제가 된다면 Blue/Green을 이용해 배포하고 모니터링한 후 문제가 있다면 빠르게 Blue 버전으로 롤백할 수 있습니다.
개발/테스트 단계에서는 Blue/Green이나 Canary를 이용해서 배포하지 않고 기본 전략인 Rolling Update를 사용하고 있습니다.


> 4. 기업에서는 instance 에 kuber 설치해서 사용 or container service 사용 중 어떤것을 선호 하나요? 어떻게 변경되어가고 있는 추세인가요?

이것은 'kubernetes를 사용해서 docker 컨테이너를 배포하는지, 아니면 kubernetes는 사용하지 않고 docker container로만 서비스 운영만 하는지'를 질문 주신 것으로 이해했습니다. 제가 본 대기업과 스타트업 모두 kubernetes를 사용해서 서비스를 운영하고 있습니다. 회사 상황에 따라서 물리 서버를 이용해 클러스터를 구성하고 사용하는 기업도 있습니다. 직접 물리 서버가 없더라도 실습에서 사용하셨던 것처럼, 클라우드 서비스 제공자(AWS, GCP 등)가 쿠버네티스 클러스터 기능을 Manual 혹은 Managed 서비스 형태로 제공하고 있기 때문에 쿠버네티스 사용성과 접근성이 높아진 것 같습니다.



혹시 제가 이해한 질문 내용이 다르거나 추가 질문이 있으시면 언제든 남겨주세요.
감사합니다. ☺️

**Reply by Tyler-devop** [reply]
안녕하세요! 먼저 자세하고 긴답변 주셔서 감사합니다! 
4번 문항에 대해 추가 질문이 있어서 문의드립니다! 

4번 질문의 본 의도는 1)EC2 같은 instance 에 kubernetes 를 직접 manual 로 설치를 해서 사용을 하는지 와 2)AWS EKS, GCP GKE 등 처럼 kubernetes service 를 사용하는지 여쭤본 것이었습니다. 운영하는데 있어 비용적인 차이가 있을것 같아서 무엇을 선호하는지 궁금해서 여쭤보았습니다!

**Reply by yoonjeong-kwon** [reply]
안녕하세요.
EKS는 Managed 쿠버네티스 서비스이기 때문에 EC2를 이용해 직접 클러스터를 구성하시는 것보다 관리가 쉽고 비용 측면에서도 저렴한 것으로 알고 있습니다.
만약 실제 현업에서 클러스터를 구성을 고려하고 계시다면, 요구사항에 따라 AWS나 GCP에서 추천하는 방법이 다를 수 있기 때문에 각 지원팀에 문의 후 도움 받으시는 것을 추천드립니다.

**Comment by yoonjeong-kwon** [comment]
@Tyler-devop 님 안녕하세요. 이후 Chapter 04를 학습 하시면서 나올 내용을 추가로 답변드립니다.

위에서 Service와 Deployment 실행 순서에 관해서 질문을 주셨습니다. 일반적으로는 Service 이름을 이용해 다른 Pod와 통신하도록 구성을 하게 되는데요. 이런 경우에는 위에서 답변드린대로 실행 순서에 영향을 받지 않습니다. 
그러나 만약 애플리케이션 컨테이너에서 쿠버네티스가 설정해주는 Service 관련 환경변수를 사용해 다른 Pod와 통신할 계획이라면, Service를 먼저 생성한 후에 Pod를 생성해야 합니다. 왜냐하면 Pod 생성 시 그 시점에 존재하는 Service 오브젝트들에 대해서만 쿠버네티스가 환경변수로 설정해주기 때문입니다. 따라서 Pod 생성 이후 만들어진 Service 관련 환경변수는 애플리케이션 컨테이너에서 참조할 수 없습니다. 이 부분은 강의자료 Ch 04. 쿠버네티스로 서비스 운영하기 - 01. Service 소개의 슬라이드 9번을 참고해주세요.

감사합니다.

**Comment by yoonjeong-kwon** [comment]
Tyler-devop님, 안녕하세요.

추가 질문 주신 것에 대한 답변을 확인하셨는지 궁금하네요 :)
그 외에 이후 학습하시면서 나올 내용에 대한 추가 답변도 올려두었습니다.
https://github.com/yoonjeong-kwon/fastcampus-kubernetes/discussions/5#discussioncomment-2559165
학습에 참고 부탁드립니다.

감사합니다.


2022년 4월 9일 (토) 오후 8:28, Tyler-devop ***@***.***>님이 작성:

**Reply by Tyler-devop** [reply]
네! 답변 모두 확인했고 질문에 대해 궁금증이 모두 해결되었습니다! 추가답변까지... 감사합니다! 
추가 문의 사항이 생기면 또 여쭤보도록 하겠습니다!
